bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -976,6 +976,47 @@ void PreProcessor::onMessage<PreProcessReplyMsg>(PreProcessReplyMsg *message) {
   handleSinglePreProcessReplyMsg(msg);
 }
 
+bool PreProcessor::checkPreProcessReplyPrerequisites(
+    SeqNum reqSeqNum, const string &cid, NodeIdType senderId, const string &batchCid, uint16_t offsetInBatch) {
+  if (myReplica_.isCollectingState()) {
+    LOG_INFO(logger(),
+             "Ignore PreProcessReplyMsg as the replica is collecting missing state from other replicas"
+                 << KVLOG(reqSeqNum, cid, batchCid, senderId, offsetInBatch));  // TODO (Efrat) params list
+    return false;
+  }
+
+  if (!myReplica_.isCurrentPrimary()) {
+    LOG_WARN(logger(),
+             "Ignore PreProcessReplyMsg as current replica is not the primary"
+                 << KVLOG(reqSeqNum, cid, batchCid, senderId, offsetInBatch));
+    return false;
+  }
+
+  if (!myReplica_.currentViewIsActive()) {
+    LOG_INFO(logger(),
+             "Ignore PreProcessReplyMsg as current view is inactive"
+                 << KVLOG(reqSeqNum, cid, batchCid, senderId, offsetInBatch));
+    return false;
+  }
+  return true;
+}
+
+bool PreProcessor::checkPreProcessBatchReplyMsgCorrectness(const PreProcessBatchReplyMsgSharedPtr &batchReply) {
+  const auto &preProcessReplyMsgs = batchReply->getPreProcessReplyMsgs();
+  NodeIdType senderId = batchReply->senderId();
+  const string batchCid = batchReply->getCid();
+  bool valid = true;
+
+  for (const auto &replyMsg : preProcessReplyMsgs) {
+    if (!checkPreProcessReplyPrerequisites(
+            replyMsg->reqSeqNum(), replyMsg->getCid(), senderId, batchCid, replyMsg->reqOffsetInBatch()) ||
+        !validateMessage(replyMsg.get())) {
+      valid = false;
```

Comment:
Add 'break' here. No need to continue verifications in case one of the messages failed to pass them.